### DISCUSSION TITLE
How to use set_position to center the window on the screen?

### DISCUSSION THREAD

**Discussion by ultimaweapon** [slint-ui/slint#3859] [Q&A] [answered]
I'm trying to center the window with the following code:

```rust
let win = MainWindow::new().with_context(|| "Cannot create main window")?;

win.window().set_position(????);
```

My question is how to specify the center position? Do I need to calculate the position manually?

**Comment by bjorn** [comment, marked as answer]
> My question is how to specify the center position? Do I need to calculate the position manually?

Note that first of all, I think it's quite unusual for applications to position their window explicitly. It is generally better left to the window manager to choose a good starting location (which might take into account existing windows, current focus, mouse position, etc.). Exception might be when you're restoring the previous state of the application, or when the window is an alert that needs to be centered for attention.

But to answer the question, yes you need to calculate the position manually. This will require using some private API to access the `winit` window and then using that `winit` API to get monitor size. See:

* https://github.com/rust-windowing/winit/issues/38#issuecomment-1513727854 for links to monitor size API.
* https://github.com/rust-windowing/winit/blob/e1edc1d6b7e42b8d7632b18dfd87f5c75b831e63/src/window.rs#L420-L436 for window centering code that wasn't merged into winit.
* https://slint.dev/releases/1.2.2/docs/rust/slint/struct.Window#impl-WinitWindowAccessor-for-Window for Slint's functions to interface with winit (which as far as I understand requires using `i-slint-backend-winit` crate).

**Reply by ultimaweapon** [reply]
Thanks.

**Reply by prof79** [reply]
That was a tough act taking several hours, at least an idea of how this might be done. The dynamic `AppWindow` structure and `winit` API changes in-between didn't make things any easier.

https://github.com/prof79/slint-center-window
